test/e2e/data/cluster-templates/docker-rke2.yaml

```
@@ -28,72 +16,153 @@ spec:
+              name: ${CLUSTER_NAME}-worker-machinetemplate
+  variables:
+    - name: rke2CNI
+      required: true
+      schema:
+        openAPIV3Schema:
+          type: string
+    - name: dockerImage
+      required: true
+      schema:
+        openAPIV3Schema:
+          type: string
+  patches:
+    - name: rke2cni
+      definitions:
+        - selector:
+            apiVersion: controlplane.cluster.x-k8s.io/v1beta1
+            kind: RKE2ControlPlaneTemplate
+            matchResources:
+              controlPlane: true
+          jsonPatches:
+            - op: add
+              path: /spec/template/spec/serverConfig/cni
+              valueFrom:
+                variable: rke2CNI
+    - name: dockerImage
+      definitions:
+        - selector:
+            apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+            kind: DockerMachineTemplate
+            matchResources:
+              controlPlane: true
+              machineDeploymentClass:
+                names:
+                  - default-worker
+          jsonPatches:
+            - op: add
+              path: /spec/template/spec/customImage
+              valueFrom: 
+                variable: dockerImage
 ---
 apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
-kind: DockerMachineTemplate
+kind: DockerClusterTemplate
 metadata:
-  name:  ${CLUSTER_NAME}-control-plane
+  name: ${CLUSTER_NAME}
   namespace: ${NAMESPACE}
 spec:
-  template:
-    spec:
-      customImage: kindest/node:${KUBERNETES_VERSION}
-      bootstrapTimeout: 15m
+  loadBalancer:
+    customHAProxyConfigTemplateRef:
+      name: ${CLUSTER_NAME}-lb-config
 ---
 apiVersion: controlplane.cluster.x-k8s.io/v1beta1
-kind: RKE2ControlPlane
+kind: RKE2ControlPlaneTemplate
 metadata:
   name: ${CLUSTER_NAME}-control-plane
   namespace: ${NAMESPACE}
 spec:
-  replicas: ${CONTROL_PLANE_MACHINE_COUNT}
-  version: ${RKE2_VERSION}
-  rolloutStrategy:
-    rollingUpdate:
-      maxSurge: 1
-    type: RollingUpdate
-  serverConfig:
-    cloudProviderName: external
-    cni: ${RKE2_CNI}
-    kubeAPIServer:
-      extraArgs:
-      - --anonymous-auth=true
-    disableComponents:
-      pluginComponents:
-       - rke2-ingress-nginx
-      kubernetesComponents:
-      - cloudController
-    etcd:
-      backupConfig:
-        scheduleCron: '*/5 * * * *'
-        retention: "1"
-  agentConfig:
-    nodeAnnotations:
-      test: "true"
-  machineTemplate:
-    infrastructureRef:
-      apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
-      kind: DockerMachineTemplate
-      name:  ${CLUSTER_NAME}-control-plane
-    nodeDrainTimeout: 30s
-  infrastructureRef:
-    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
-    kind: DockerMachineTemplate
-    name:  ${CLUSTER_NAME}-control-plane
-  nodeDrainTimeout: 30s  
+  template:
+    spec:
+      rolloutStrategy:
+        rollingUpdate:
+          maxSurge: 1
+        type: RollingUpdate
+      serverConfig:
+        cloudProviderName: external
+        cni: 
```

```suggestion
        cni: None
```

to make sure the template is valid, even if a variable replaces it

---

this has to be removed, I'm adding variables with `add` operation

---

There is a replace operation to make it work. [CNI](https://github.com/rancher/cluster-api-provider-rke2/blob/a2e35699da1cd6eb0e0cda66431933f3b602499e/controlplane/api/v1beta1/rke2controlplane_types.go#L357) is a enum resource, it shouldn’t accept empty values.

---

it's optional in the CRD https://github.com/rancher/cluster-api-provider-rke2/blob/a2e35699da1cd6eb0e0cda66431933f3b602499e/controlplane/api/v1beta1/rke2controlplane_types.go#L159

---

In this case please remove the field.

---

and that makes me think we should make it required in CAPRKE2